Overview
-------------------------------

Input for SEVtras is a cell-by-gene matrix. In the case of scRNA-seq dataset using 10X Genomics, we used `raw_feature_bc_matrix` directory generated by Cell Ranger as input in the first step. Output of SEVtras consists of the score of sEV signals and classification for each droplet. Such sEV information will be used for downstream analysis and as basis for the construction of the sEV secretion activity index (ESAI) for different cell types. 

We implemented four functions for sEV recognizing and functional analyses. In SEVtras, `sEV_recognizer` recognizes sEV-containing droplets in the raw scRNA-seq data; `ESAI_calculator` calculates sEV secretion activity for samples and deconvolves these droplets to their original cell type and estimited corresponding sEV secretion activity, which may be related to tumor tumor malignancy, invasion, metastasis and other disease progression. 

You can freely use SEVtras to explore sEV heterogeneity at single droplet, characterize cell type dynamics in light of sEV activity and unlock diagnostic potential of sEVs in concert with cells. 
